### Table of Contents

- [Introduction](#intro)
- [Usage](#use)
- [Arguments](#args)
- [Output explanation](#out)

## <a name="intro"></a>Introduction

`sqanti_sim.py eval` is the final step of the SQANTI-SIM pipeline. In this step SQANTI3 is run and a report is generated by SQANTI-SIM to evaluate the performance of the pipeline employed to identify the transcripts. This mode takes your long read-defined transcriptome, the modified reference annotation and the reference genome used in the steps before to evaluate the performance of your pipeline.

In this step you are able to provide **orthogonal** data such as CAGE Peak and support with short read (that maybe you used to in your reconstruction pipeline) to generate also some metrics in the report regarding this information.

## <a name="use"></a>Usage

SQANTI-SIM *eval* mode usage:

```
sqanti_sim.py eval [-h] --isoforms ISOFORMS --gtf GTF --genome GENOME
                          -i TRANS_INDEX [-o OUTPUT] [-d DIR]
                          [--short_reads SHORT_READS] [--cage_peak CAGE_PEAK]
                          [--min_support MIN_SUPPORT] [-k CORES]
```

With the `--help` option you can display a full description of the arguments:

```
sqanti_sim.py eval parse options

optional arguments:
  -h, --help            show this help message and exit
  --isoforms ISOFORMS   GTF with trancriptome reconstructed with your pipeline
  --gtf GTF             Reference annotation in GTF format
  --genome GENOME       Reference genome FASTA
  -i TRANS_INDEX, --trans_index TRANS_INDEX
                        File with transcript information generated with
                        SQANTI-SIM
  -o OUTPUT, --output OUTPUT
                        Prefix for output files
  -d DIR, --dir DIR     Directory for output files (default: .)
  --short_reads SHORT_READS
                        File Of File Names (fofn, space separated) with paths
                        to FASTA or FASTQ from Short-Read RNA-Seq. If
                        expression or coverage files are not provided,
                        Kallisto (just for pair-end data) and STAR,
                        respectively, will be run to calculate them.
  --cage_peak CAGE_PEAK
                        FANTOM5 Cage Peak (BED format, optional)
  --min_support MIN_SUPPORT
                        Minimum number of supporting reads for an isoform
  -k CORES, --cores CORES
                        Number of cores to run in parallel
```

Running the tool with the minimum input to simulate PacBio reads will look as follows:

```
(SQANTI-SIM.env)$ python sqanti_sim.py eval \
			--isoforms long_read_transcriptome.gtf \
			--trans_index prefix_index.tsv \
			--gtf modified_reference_annotation.gtf \
			--genome reference_genome.fasta \
```

# TODO

## <a name="args"></a>Arguments detailed explanation

### Required input

These are the minimal parameters you will need to run `sqanti_sim.py eval`:

- **Transcript index** file (`-i`): This file is the *prefix_index.tsv* file generated in the previus *preparatory* step. New columns and information will be add to this file, so you should track it and use the most updated version of this file in the next steps of the SQANTI-SIM pipeline.
- **Long-read transcriptome** in GTF format (`--isoforms`): The isoforms identified with your transcript reconstruction pipeline. The transcripts models must be in GTF format.
- **Reference annotation** in GTF format (`--gtf`): This file is the modified GTF reference annotation generated in the *preparatory* step, that you should have used in your transcript reconstruction pipeline.
- **Reference genome** in FASTA format (`--genome`): This is the reference genome in FASTA format.

### Optional input

- **Short reads support** (`--short_reads`): File Of File Names (fofn, space separated) with paths to FASTA or FASTQ from Short-Read RNA-Seq.
- **CAGE Peak data** (`--cage_peak`): FANTOM5 Cage Peak (BED format, optional)
- **Output prefix**(`-o`): The output prefix for the index file. SQANTI-SIM will use "sqanti_sim" as default prefix.
- **Output directory**(`-d`): Output directory for output files. SQANTI-SIM will use the directory where the script was run as the default output directory.
- **Supporting reads** (`--min_support`): Minimum number of supporting reads for an isoform to be identified as a new isoform by your transcript reconstruction pipeline. This parameter don't affec the results, its just to generate some metrics in the report according to your pipeline limitations/thresholds.
- **Parallelization**(`-k`): Number of cores to run in parallel. Most of the SQANTI-SIM modes have code chunks that can be run in parallel, however the default option is to run SQANTI-SIM in one single thread.

## <a name="out"></a>Output explanation

#### prefix_SQANTI_SIM_report.html

This is the main output file of the SQANTI-SIM pipeline. This HTML report picks up all the performance metrics of your transcript identification pipeline.

#### SQANTI3 output files

This is a directory with all the SQANTI3 output. A detailed explanation of its output can be found in the [SQANTI3 wiki]().



