## What is SQANTI-SIM?

*SQANTI-SIM* is a simulator of controlled novelty and degradation of transcripts sequenced by long reads. It is a wrapper tool for RNA-Seq long-reads simulators such as [IsoSeqSim](https://github.com/yunhaowang/IsoSeqSim) and [NanoSim](https://github.com/bcgsc/NanoSim) (formerly Trans-NanoSim) to simulate transcripts based on the SQANTI3 structural categories ([publication](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5848618/) and [code repository](https://github.com/ConesaLab/SQANTI3)).

The aim of SQANTI-SIM is to simulate novel and degradated transcripts in a controlled way using as ground-truth real transcripts in the reference annotation that fit the description of the different SQANTI3 structural categories. SQANTI-SIM simulates reads from real annotated transcripts and modifies the GTF reference annotation pretending that the original transcripts are not annotated (novel). This allows to have a robust ground-truth and also to use cross-sectional data (Short-read RNA-seq data, CAGE Peak data, polyA motif...) that can be use to support the discoverage and reconstruction of this "novel" transcripts that were simulated.

![workflow](https://github.com/jorgemt98/SQANTI-SIM/blob/main/docs/sqantisim_workflow.png)

SQANTI-SIM transcript classification is based on the structural categories defined by SQANTI3. A detailed explanation of the SQANTI3 structural categories can be found in the [SQANTI3 wiki](). As a brief summary, degradated transcripts which have fewer 5' start and/or 3' end exons are classified as an *Incomplete Splice Match* (ISM). On the other hand, novel isoforms are mainly classified as *Novel In Catalog* (NIC) or *Novel Not in Catalog* (NNC). NIC isoforms use a new combination of known donor/acceptor sites, while NNC isoforms have at least one donor or acceptor site that is not annotated. With SQANTI-SIM we are also able to simulate all the rest structural categories defined by SQANTI3.

### How it works?

The SQANTI-SIM pipeline consists mainly in 3 different steps: (i) simulate data, (ii) reconstruct the transcriptome with your pipeline and (iii) evaluate the performance of the transcript reconstruction pipeline. `sqanti_sim.py` is a wrapper script with modules that are assumed to be run in the following order: classif, preparatory, sim and eval. The 3 first modules will simulate the data and the last module would generate the evalution report.

1. **Transcript classification in SQANTI3 structural category**

This is the first step of the workflow. In this step the SQANTI3 structural category decision tree is used to classify all the annotated transcripts from a given GTF reference annotation in its potential SQANTI3 structural categories, that is assigning the query transcript a SQANTI3 structural category assuming it cannot match as a FSM with itself. This will generate an index tab-separated file with all the structural annotation of each transcript from the reference GTF and its potential SQANTI3 structural category.

Information about how to run this step can be found at [*SQANTI-SIM: classif*]().

2. **Simulation experiment design**

This steps takes the GTF reference annotation and the index file generated before and modifies the GTF according to the structural categories you want to simulate. It also generates the expression distribution and the TPM values for each transcript that will be simulated. Transcript TPM values can be generated by 3 different modes: using a uniform distribution for the count values, using 2 custom  negative binomial distribution (one for the novel transcripts and the other for the known) and finally it can simulate the TPM values from the quantification of the expression levels of a transcriptome mapping with Minimap2 against the reference transcriptome.

3. **sim step**


4. **eval step**



