---
title: "SQANTISIM OUTPUT COMPARISON REPORT"
author: "`r paste0('  ','Author', ': ',Sys.getenv('USERNAME'))`"
date: "`r paste0('  ','Date', ': ', format(Sys.time(), '%d %B %Y'))`"
output:
  html_document:
    number_sections: false
    toc: true
    toc_float: true
    toc_depth: 3
    theme: spacelab
---

<!-- Basic setting -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, cache = FALSE, echo = FALSE)
```

<!-- Report generation -->

## Summary

```{r, results='asis'}
n.sim.trans <- nrow(res.full$data.known) +  nrow(res.full$data.novel)
n.known.trans <- nrow(res.full$data.known)
n.novel.trans <- nrow(res.full$data.novel)
n.class.trans <- nrow(data.query)

cat('\n\n**Number of simulated transcripts**: ', n.sim.trans, '\n\n')
cat('\n\n**Number of known transcripts**: ', n.known.trans, '\n\n')
cat('\n\n**Number of novel transcripts**: ', n.novel.trans, '\n\n')
cat('\n\n**Number of reconstructed transcripts**: ', n.class.trans, '\n\n')
```

## Simulated counts distribution

```{r}
p1
```

## Isoform classification distribution of reconstructed transcripts

```{r}
p2
```

## Evaluation of simulation {.tabset}

Simulated transcript were grouped in know simulated transcripts which appear in the reference annotation GTF and novel simulated transcripts that were deleted from the annotation. The following metrics were computed:

- **Number of isoforms simulated**: Total number of simulated isoforms
- **True Positive isoforms (TP)**: Isoforms that match junctions with a simulated transcript and its TSS and TTS is within the 50bp threshold
- **Partial True Positive isoforms (PTP)**: Isoforms that match junctions but the TSS and TTS are beyond the threshold
- **False Negative isoforms (FN) **: Isoforms that were simulated but not detected
- **False Positive (FP)**: Transcript that were detected but not simulated
- **Sensitivity**: TP/Number of simulated transcripts
- **Precision**: TP/Detected isoforms
- **F-score**: Harmonic mean of precision and sensitivity
- **False Discovery Rate**: (FP+PTP)/Detected isoforms
- **Positive Detection Rate**: (TP + PTP)/Number of simulated transcripts
- **False Detection Rate**: FP/Detected isoforms

### Full results

#### Stats for known transcripts

```{r}
t1
```

```{r, results = 'hide'}
radar.palette <- c("global"='hotpink', "FSM"="#6BAED6", "ISM"="#FC8D59", "NIC"="#78C679", 
                "NNC"="#EE6A50", "Genic\nGenomic"="#969696", "Antisense"="#66C2A4", "Fusion"="goldenrod1",
                "Intergenic" = "darksalmon", "Genic\nIntron"="#41B6C4")
known.metrics.t <- as.data.frame(t(as.matrix(res.full$known.metrics)))
known.metrics.t[is.na(known.metrics.t)] <- NA
known.metrics.t <-known.metrics.t[,c("Sensitivity", "Precision", "F-score", "False_Discovery_Rate", "Positive_Detection_Rate", "False_Detection_Rate")]
known.metrics.t <- rbind(1, 0, known.metrics.t)

par(mar=c(1,1,1,1), mfrow=c(2,ceiling((nrow(known.metrics.t)-2)/2)))
for (i in 1:(nrow(known.metrics.t)-2)){
  radarchart(
  known.metrics.t[c(1, 2, i+2),], axistype = 1,
  pcol = radar.palette[[rownames(known.metrics.t)[i+2]]], pfcol = scales::alpha(radar.palette[[rownames(known.metrics.t)[i+2]]], 0.5), plwd = 2, plty = 1,
  # Customize the grid
  cglcol = "grey", cglty = 1, cglwd = 0.8,
  # Customize the axis
  axislabcol = "grey", 
  # Variable labels
  vlcex = 0.7, vlabels = colnames(known.metrics.t[c(1, 2, i+2),]),
  caxislabels = c(0, 0.25, 0.5, 0.75, 1), title = rownames(known.metrics.t)[i+2]
)
}
```

#### Stats for novel transcripts

```{r}
t2
```

```{r, results = 'hide'}
radar.palette <- c("global"='hotpink', "FSM"="#6BAED6", "ISM"="#FC8D59", "NIC"="#78C679", 
                "NNC"="#EE6A50", "Genic\nGenomic"="#969696", "Antisense"="#66C2A4", "Fusion"="goldenrod1",
                "Intergenic" = "darksalmon", "Genic\nIntron"="#41B6C4")
novel.metrics.t <- as.data.frame(t(as.matrix(res.full$novel.metrics)))
novel.metrics.t[is.na(novel.metrics.t)] <- NA
novel.metrics.t <-novel.metrics.t[,c("Sensitivity", "Precision", "F-score", "False_Discovery_Rate", "Positive_Detection_Rate", "False_Detection_Rate")]
novel.metrics.t <- rbind(1, 0, novel.metrics.t)

par(mar=c(1,1,1,1), mfrow=c(2,ceiling((nrow(novel.metrics.t)-2)/2)))
for (i in 1:(nrow(novel.metrics.t)-2)){
  radarchart(
  novel.metrics.t[c(1, 2, i+2),], axistype = 1,
  pcol = radar.palette[[rownames(novel.metrics.t)[i+2]]], pfcol = scales::alpha(radar.palette[[rownames(novel.metrics.t)[i+2]]], 0.5), plwd = 2, plty = 1,
  # Customize the grid
  cglcol = "grey", cglty = 1, cglwd = 0.8,
  # Customize the axis
  axislabcol = "grey", 
  # Variable labels
  vlcex = 0.7, vlabels = colnames(novel.metrics.t[c(1, 2, i+2),]),
  caxislabels = c(0, 0.25, 0.5, 0.75, 1), title = rownames(novel.metrics.t)[i+2]
)
}
```

#### Number of exons

```{r}
p3
```

#### Canonical Junctions

```{r}
p4
```

```{r, results='asis'}
if ('within_cage_peak' %in% colnames(data.index)){
  cat('\n\n#### CAGE Peak Support\n\n')
  print(p5)
  print(p6)
}

if ('min_cov' %in% colnames(data.index)){
  cat('\n\n#### Short Reads Support\n\n')
  print(p7)
  print(p8)
}
```

### Above min. supp. results

#### Stats for known transcripts

```{r}
t1.min
```

```{r, results = 'hide'}
radar.palette <- c("global"='hotpink', "FSM"="#6BAED6", "ISM"="#FC8D59", "NIC"="#78C679", 
                "NNC"="#EE6A50", "Genic\nGenomic"="#969696", "Antisense"="#66C2A4", "Fusion"="goldenrod1",
                "Intergenic" = "darksalmon", "Genic\nIntron"="#41B6C4")
known.metrics.t <- as.data.frame(t(as.matrix(res.min$known.metrics)))
known.metrics.t[is.na(known.metrics.t)] <- NA
known.metrics.t <-known.metrics.t[,c("Sensitivity", "Precision", "F-score", "False_Discovery_Rate", "Positive_Detection_Rate", "False_Detection_Rate")]
known.metrics.t <- rbind(1, 0, known.metrics.t)

par(mar=c(1,1,1,1), mfrow=c(2,ceiling((nrow(known.metrics.t)-2)/2)))
for (i in 1:(nrow(known.metrics.t)-2)){
  radarchart(
  known.metrics.t[c(1, 2, i+2),], axistype = 1,
  pcol = radar.palette[[rownames(known.metrics.t)[i+2]]], pfcol = scales::alpha(radar.palette[[rownames(known.metrics.t)[i+2]]], 0.5), plwd = 2, plty = 1,
  # Customize the grid
  cglcol = "grey", cglty = 1, cglwd = 0.8,
  # Customize the axis
  axislabcol = "grey", 
  # Variable labels
  vlcex = 0.7, vlabels = colnames(known.metrics.t[c(1, 2, i+2),]),
  caxislabels = c(0, 0.25, 0.5, 0.75, 1), title = rownames(known.metrics.t)[i+2]
)
}
```

#### Stats for novel transcripts

```{r}
t2.min
```

```{r, results = 'hide'}
radar.palette <- c("global"='hotpink', "FSM"="#6BAED6", "ISM"="#FC8D59", "NIC"="#78C679", 
                "NNC"="#EE6A50", "Genic\nGenomic"="#969696", "Antisense"="#66C2A4", "Fusion"="goldenrod1",
                "Intergenic" = "darksalmon", "Genic\nIntron"="#41B6C4")
novel.metrics.t <- as.data.frame(t(as.matrix(res.min$novel.metrics)))
novel.metrics.t[is.na(novel.metrics.t)] <- NA
novel.metrics.t <-novel.metrics.t[,c("Sensitivity", "Precision", "F-score", "False_Discovery_Rate", "Positive_Detection_Rate", "False_Detection_Rate")]
novel.metrics.t <- rbind(1, 0, novel.metrics.t)

par(mar=c(1,1,1,1), mfrow=c(2,ceiling((nrow(novel.metrics.t)-2)/2)))
for (i in 1:(nrow(novel.metrics.t)-2)){
  radarchart(
  novel.metrics.t[c(1, 2, i+2),], axistype = 1,
  pcol = radar.palette[[rownames(novel.metrics.t)[i+2]]], pfcol = scales::alpha(radar.palette[[rownames(novel.metrics.t)[i+2]]], 0.5), plwd = 2, plty = 1,
  # Customize the grid
  cglcol = "grey", cglty = 1, cglwd = 0.8,
  # Customize the axis
  axislabcol = "grey", 
  # Variable labels
  vlcex = 0.7, vlabels = colnames(novel.metrics.t[c(1, 2, i+2),]),
  caxislabels = c(0, 0.25, 0.5, 0.75, 1), title = rownames(novel.metrics.t)[i+2]
)
}
```

#### Number of exons

```{r}
p3.min
```

#### Canonical Junctions

```{r}
p4.min
```

```{r, results='asis'}
if ('within_cage_peak' %in% colnames(data.index)){
  cat('\n\n#### CAGE Peak Support\n\n')
  print(p5.min)
  print(p6.min)
}

if ('min_cov' %in% colnames(data.index)){
  cat('\n\n#### Short Reads Support\n\n')
  print(p7.min)
  print(p8.min)
}
```
