---
title: "SQANTI-SIM OUTPUT COMPARISON REPORT"
author: "`r paste0('  ','Author', ': ',Sys.getenv('USERNAME'))`"
date: "`r paste0('  ','Date', ': ', format(Sys.time(), '%d %B %Y'))`"
output:
  html_document:
    number_sections: false
    toc: true
    toc_float: true
    toc_depth: 3
    theme: spacelab
---

<!-- Basic setting -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, cache = FALSE, echo = FALSE)
```

<!-- Report generation -->

## Summary

```{r, results='asis'}
n.sim.trans <- nrow(data.known) +  nrow(data.novel)
n.known.trans <- nrow(data.known)
n.novel.trans <- nrow(data.novel)
n.class.trans <- nrow(data.query)

cat('\n\n**Number of simulated transcripts**: ', n.sim.trans, '\n\n')
cat('\n\n**Number of known transcripts**: ', n.known.trans, '\n\n')
cat('\n\n**Number of novel transcripts**: ', n.novel.trans, '\n\n')
cat('\n\n**Number of reconstructed transcripts**: ', n.class.trans, '\n\n')
```

## Simulated counts distribution

```{r}
p1
```

## Isoform classification distribution of reconstructed transcripts

```{r}
p2
```

## Evaluation of simulation {.tabset}

Simulated transcript were grouped in know simulated transcripts which appear in the reference annotation GTF and novel simulated transcripts that were deleted from the annotation. The following metrics were computed:

- **Number of isoforms simulated**: Total number of simulated isoforms
- **True Positive isoforms (TP)**: Isoforms that match junctions with a simulated transcript and its TSS and TTS is within the thresholds
- **Partial True Positive isoforms (PTP)**: Isoforms that match junctions but the TSS and TTS are beyond the threshold
- **False Negative isoforms (FN) **: Isoforms that were simulated but not detected
- **False Positive (FP)**: Transcript that were detected but not simulated
- **Sensitivity**: TP/Number of simulated transcripts
- **Precision**: TP/Detected isoforms
- **F-score**: Harmonic mean of precision and sensitivity
- **False Discovery Rate**: (FP+PTP)/Detected isoforms
- **Positive Detection Rate**: (TP + PTP)/Number of simulated transcripts
- **False Detection Rate**: FP/Detected isoforms

### Stats for known transcripts

```{r}
t1
```

### Stats for novel transcripts

```{r}
t2
```

#### Novel TP vs FN

```{r}
p4
```

#### Sensitivity deppending on the number of annotated transcripts per gene

```{r}
p3.1
p3.2
```